klippy/extras/frame_expansion_compensation.py

```diff
@@ -0,0 +1,195 @@
+
+# Compensates for thermal expansion induced toolhead Z movement in real-time
+# using a frame-coupled temperature probe.
+
+KELVIN_TO_CELSIUS = -273.15
+STD_TEMP = 25. # reference temperature for frame measurement
+
+class FrameExpansionCompensator:
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        self.gcode = self.printer.lookup_object('gcode')
+
+        # Get config parameters, convert to SI units where necessary
+        self.coeff = config.getfloat('coeff', minval=0., maxval=100.,)/1E6
+        self.frame_z = config.getfloat('frame_z_length', minval=0.,)/1E3
+        self.gantry_factor = config.getfloat('gantry_factor', default=1.0)
+        self.min_temp = config.getfloat('min_temp', minval=KELVIN_TO_CELSIUS)
+        self.max_temp = config.getfloat('max_temp', above=self.min_temp)
+        self.z_stepper_name = config.get('z_stepper')
+        self.max_comp_z = config.getfloat('max_comp_z', 0.)
+        self.max_offset = config.getfloat('max_z_offset', 99999999.)
+
+        # Register printer events
+        self.printer.register_event_handler("klippy:ready",
+                                            self.handle_ready)
+        self.printer.register_event_handler("homing:homing_move_end",
+                                            self.handle_homing_move_end)
+
+        # Setup temperature sensor
+        self.last_temp = 0.
+        self.measured_min = 99999999.
+        self.measured_max = 0.
+        self.last_home_temp = 0.
+        pheaters = self.printer.load_object(config, 'heaters')
+        self.sensor = pheaters.setup_sensor(config)
+        self.sensor.setup_minmax(self.min_temp, self.max_temp)
+        self.sensor.setup_callback(self.temperature_callback)
+        pheaters.register_sensor(config, self)
+
+        # Z offset transformation
+        self.z_drift_offset = 0.
+        self.comp_enable = True
+        self.last_position = [0., 0., 0., 0.]
+        self.next_transform = None
+        self.comp_state = 'not homed'
+
+        # Register gcode commands
+        self.gcode.register_command('SET_FRAME_COMP',
+                                    self.cmd_SET_FRAME_COMP,
+                                    desc=self.cmd_SET_FRAME_COMP_help)
+        self.gcode.register_command('QUERY_FRAME_COMP',
+                                    self.cmd_QUERY_FRAME_COMP,
+                                    desc=self.cmd_QUERY_FRAME_COMP_help)
+
+    def handle_ready(self):
+        'Called when Klipper successfully connected, ready.'
+        self.toolhead = self.printer.lookup_object('toolhead')
+        gcode_move = self.printer.lookup_object('gcode_move')
+
+        # Z stepper config check
+        kin = self.printer.lookup_object('toolhead').get_kinematics()
+        steppers = [s.get_name() for s in kin.get_steppers()]
+        if not self.z_stepper_name in steppers:
+            raise self.printer.config_error(
+                "'%s' is not a valid stepper" % self.z_stepper_name)
```

Comment:
It's not valid to raise an error in a "klippy:ready" event handler (see docs/Code_Overview.md).  You probably want the "klippy:connect" event handler.

Reply:
> 
> 
> Thanks.
> 
> I guess my main comment is that the configuration of this module seems more complicated than I think it could be. Is there any reason we can't simplify this to:
> 
> ```
> [frame_expansion_compensation]
> #adjustment:
> #   The amount of Z adjustment (in millimeters per degree Celsius) to apply.
> #sensor_type:
> #sensor_pin:
> #min_temp:
> #max_temp:
> #gcode_id:
> #   See the "heater_generic" and "extruder" sections for the definition of these
> #   parameters.
> ```
> 
> If I understand the code correctly, it seems to apply a Z offset based on the temperature difference since the last Z home. If so, I think both the config and code could be simplified.
> 
> See some additional comments below.
> 
> -Kevin

Thanks for the review, Kevin.

In principal I agree, and would like to simplify both before finalizing.

I would like to keep the `coefficient` config parameter name, as this is consistent with the technical term, "coefficient of linear expansion". As well, the length of the frame member is a necessary input as the change is length is proportional to the total length: `l_total = l_ref + l_ref * (coeff * (t_current - t_ref)`

`gantry_factor` is a kind of stand-in for broader compatibility I'd like to introduce for top mounted fixed gantry/moving bed machine down the line. Hard-coding the 0.5 factor necessary for Voron V2 would make it specific to that machine, so until I can think of a nice way to take care of as many use-cases as possible I would like to keep this factor accessible in the config.

Is there any reason to omit the optional limiting safety parameters? Perhaps I could hard-code some safety checks in this regard. If the `min_temp` parameter is defined, will the ADC range error out in time to avoid crashing the head into the bed while trying to e.g. correct for thermal contraction of suddenly dropping to -80C?